File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        return preorder(root);
    }
    public List<Integer> preorder(TreeNode root) {
        if (root == null) return new ArrayList<>();
        List<Integer> result = new ArrayList<>();
        result.add(root.val);
        List<Integer> leftResult = preorder(root.left);
        List<Integer> rightResult = preorder(root.right);
        result.addAll(leftResult);
        result.addAll(rightResult);
        return result;
    }
}

File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack = []
        curr = root
        result = []
        while curr != None or stack:
            while curr != None:
                stack.append(curr)
                result.append(curr.val)
                curr = curr.left
            curr = stack.pop()
            curr = curr.right
        return result